Docs/personalization reference guide (#19125)
gatsbyjs/gatsby: docs/docs/personalization.md

---
title: Personalization
---

## What is Personalization?

- Have you ever landed on a website that greeted you by name?

> - Did you ever search for a book and realized your suggested results based on your previous purchases?

Reviewer: ```suggestion
- Did you ever search for a book and realize the suggested results were based on your previous purchases?
```

These are examples of personalization. Personalization adjusts the content of a website, and creates a custom experience for users. With personalization, one website can offer a variety of experiences to different visitors.

> ## Why is Personalization Important?

Reviewer 2: Updating to follow the Gatsby Style Guide for sentence case: [Gatsby Style Guide](https://www.gatsbyjs.org/contributing/gatsby-style-guide/)
```suggestion
## Why is personalization important?
```

> If you are [creating an e-commerce site](/tutorial/ecommerce-tutorial/), personalization is key in converting your visitors into buyers. Online merchants use personalization as a way to suggest relevant items to a shopper.

Reviewer 2: I'd recommend sending them to the e-commerce section, as there is more than one approach and the tutorial focuses on Stripe only.
```suggestion
If you are [creating an e-commerce site](/docs/building-an-e-commerce-site/), personalization is key in converting your visitors into buyers. For example, online merchants use personalization as a way to suggest relevant items to a shopper.
```

> Outside of e-commerce, personalization is important for other types of websites, like e-learning. Personalization, if implemented strategically, can help educational sites deliver more targeted learning experiences. For example, if a learner is struggling with the learning the basics of JavaScript, the website can give the learner additional JavaScript practice problems.

Reviewer: ```suggestion
Outside of e-commerce, personalization is important for other types of websites, like e-learning. Personalization, if implemented strategically, can help educational sites deliver more targeted learning experiences. For example, if a learner is struggling with learning the basics of JavaScript, the website can give the learner additional JavaScript practice problems.
```

> ## Using Personalization with Gatsby

Reviewer 2: ```suggestion
## Using personalization with Gatsby
```

> Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by [React Hydration](/blog/2018-10-15-beyond-static-intro/#hydration). During build time Gatsby invokes the React server-side DOM to generate content.

Reviewer 2: This is a great start, since personalization is such an under-documented/supported area of Gatsby. Here's a suggestion on how to restructure it to emphasize what Gatsby _can_ do while being realistic about the situation:
```suggestion
Personalization involves rendering dynamic content to a user. The way Gatsby generates static content at [build](/docs/glossary#build) time and [hydrates into a React web application](/blog/2018-10-15-beyond-static-intro/#hydration) both limits what can be done to render dynamic content at build time, and creates opportunities to enhance content with client-side JavaScript.

One possible way to go about personalization is through lazy-loading of content with client-side JavaScript, by requesting specific content from a web/cloud service or [lambda function](https://docs.aws.amazon.com/lambda/latest/dg/welcome.html) and displaying it to a user when they view a page. A benefit of this approach would be the ability to access dynamic content per user with your Gatsby site, with minimal infrastructure changes. A drawback would be that it relies on client-side JavaScript to render content, which could change from the initial server-rendered view and cause a flash of changed content.
```

When we have a specific doc on hydration, we could cover this use case. cc @gillkyle.

Reviewer 3: Should wenn not only show drawbacks but also mention the benefits: if a wireframe is loaded fast from prebuild files then the user get an fast feedback. 

Facebook make it also this way and they show a placeholder while loading user content

> Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by [React Hydration](/blog/2018-10-15-beyond-static-intro/#hydration). During build time Gatsby invokes the React server-side DOM to generate content.

Author: Hydration in relation to personalization was particularly challenging to describe. That being said, reading @gillkyle's [adding rehydration doc and updates to building apps overview PR](https://github.com/gatsbyjs/gatsby/pull/19214) was very helpful in helping me form a mental model of how hydration and rehydration works in Gatsby. 

If I'm missing any details about react hydration, or if my understanding of how Gatsby handles dynamic data is a little off, please let know!

> One possible way to go about personalization is through edge side includes (ESI). Edge side includes is a proposed web standard that allows developers to load dynamic content onto a webpage. The idea is you can cache most of your webpage except for the personalized portion of it. In the example of the e-commerce site, you will cache everything but the recommendation portion of the site. ESI can be [implemented in a service worker and you can learn more about it here](https://blog.cloudflare.com/edge-side-includes-with-cloudflare-workers/).

Reviewer 2: ```suggestion

### Edge side includes

Another potential approach to personalization in Gatsby would be edge side includes (ESI). Edge side includes are a proposed web standard to allow developers to load dynamic content onto a webpage. The idea is you can cache most of your webpage except for the personalized portion of it. In the example of the e-commerce site, you could cache everything but the recommendation portion of the site. ESI can be [implemented in a service worker and you can learn more about it here](https://blog.cloudflare.com/edge-side-includes-with-cloudflare-workers/). 

A benefit to this approach would be better page rendering performance on less powerful devices, a common goal of creating a Gatsby site. Drawbacks would include having to implement an additional layer of technology–thus adding complexity to the system–and the ongoing management/support requirements. Managing an API endpoint for client-side content updates is a much more common scenario in a Gatsby application.
```

> An alternative way to personalize your Gatsby site is to use the search engine Algolia. Algolia offers personalization by using inputs like Google Analytics, and the geography of the user when displaying a list of search results. Search results are also weighted on past behavior of other site visitors. You can get started using Algolia with Gatsby by downloading the [Gatsty Algolia plug-in](/docs/adding-search-with-algolia/).

Reviewer 2: ```suggestion
An alternative way to personalize your Gatsby site is to use the search engine Algolia on the client-side. Algolia offers [personalization](https://www.algolia.com/doc/guides/getting-insights-and-analytics/personalization/) by using inputs like Google Analytics, and the geography of the user when displaying a list of search results. Search results are also weighted on past behavior of other site visitors. You can get started using Algolia with Gatsby by downloading the [Gatsty Algolia plug-in](/docs/adding-search-with-algolia/), which will help you interface with their APIs and servers.
```

## Other Resources

> [W3C ESI Docs](https://www.w3.org/TR/esi-lang)

Reviewer 2: ```suggestion
- [Beyond Static: hydration in Gatsby](/blog/2018-10-15-beyond-static-intro/#hydration)
- [W3C Edge Side Includes Docs](https://www.w3.org/TR/esi-lang)
- [Personalization with Algolia](https://www.algolia.com/doc/guides/getting-insights-and-analytics/personalization/)
```

---
title: Personalization
---

## What is Personalization?

- Have you ever landed on a website that greeted you by name?

- Did you ever search for a book and realized your suggested results based on your previous purchases?

These are examples of personalization. Personalization adjusts the content of a website, and creates a custom experience for users. With personalization, one website can offer a variety of experiences to different visitors.

## Why is Personalization Important?

> If you are [creating an e-commerce site](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/), personalization is key in converting your visitors into buyers. Online merchants use personalization as a way to suggest relevant items to a shopper.

Author: Gotcha, will take out the domain for all of the internal links.

Outside of e-commerce, personalization is important for other types of websites, like e-learning. Personalization, if implemented strategically, can help educational sites deliver more targeted learning experiences. For example, if a learner is struggling with the learning the basics of JavaScript, the website can give the learner additional JavaScript practice problems.

## Using Personalization with Gatsby

> Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by React Hydration. During build time Gatsby invokes the React server-side DOMS to generate content. You can [learn more about it here](https://www.gatsbyjs.org/blog/2018-10-15-beyond-static-intro/#hydration).

Reviewer 3: not shure - small s on plural? or remove the s here?

```suggestion
Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by React Hydration. During build time Gatsby invokes the React server-side DOMs to generate content. You can [learn more about it here](https://www.gatsbyjs.org/blog/2018-10-15-beyond-static-intro/#hydration).
```

> Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by React Hydration. During build time Gatsby invokes the React server-side DOMS to generate content. You can [learn more about it here](https://www.gatsbyjs.org/blog/2018-10-15-beyond-static-intro/#hydration).

Reviewer 3: maybe better link text

```suggestion
Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by React Hydration. During build time Gatsby invokes the React server-side DOMS to generate content. In this blog post you can learn more about [React Hydration](https://www.gatsbyjs.org/blog/2018-10-15-beyond-static-intro/#hydration).
```

> Personalization contains dynamic content. Personalization is possible in Gatsby because Gatsby can enhance static content by React Hydration. During build time Gatsby invokes the React server-side DOMS to generate content. You can [learn more about it here](https://www.gatsbyjs.org/blog/2018-10-15-beyond-static-intro/#hydration).

Author: Good eye. DOMs is incorrect. The sentence is only referring to one DOM, so the 's' should be deleted.

> One possible way to go about personalization is through ESI- edge side includes. Edge side includes is a proposed web standard that allows developers to load dynamic content onto a webpage. The idea is you can cache most of your webpage except for the personalized portion of it. In the example of the ecommerce site, you will cache everything but the recommendation portion of the site. ESI can be implemented in a service worker and you can [learn more about it here](https://blog.cloudflare.com/edge-side-includes-with-cloudflare-workers/).

Reviewer 3: - add `-` to ecommerce
- put ESI in brackets

```suggestion
One possible way to go about personalization is through edge side includes (ESI). Edge side includes is a proposed web standard that allows developers to load dynamic content onto a webpage. The idea is you can cache most of your webpage except for the personalized portion of it. In the example of the e-commerce site, you will cache everything but the recommendation portion of the site. ESI can be implemented in a service worker and you can [learn more about it here](https://blog.cloudflare.com/edge-side-includes-with-cloudflare-workers/).
```

> An alternative way to personalize your Gatsby site is to use the search engine Algoria. You can find the Gatsby Algoria plug-in here. Algoria offers personalization by using inputs like Google Analytics, and the geography of the user when displaying a list of search results. Search results are also weighted on past behavior of other site visitors.

Reviewer 3: - typo Algoria -> Algolia
- i know i am not correct in spelling but i like to remove the `You can find ... here`

```suggestion
An alternative way to personalize your Gatsby site is to use the search engine Algolia. (See the reference guide [Adding Search with Algolia](/docs/adding-search-with-algolia/)) Algolia offers personalization by using inputs like Google Analytics, and the geography of the user when displaying a list of search results. Search results are also weighted on past behavior of other site visitors. 
```

## Other Resources

> [Gatsby Algolia Plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-algolia/)

Reviewer 3: ```suggestion
[Gatsby Algolia Plugin](/packages/gatsby-plugin-algolia/)
```
